src/TropicalGeometry/hypersurface.jl

```
@@ -0,0 +1,258 @@
+# If M=min, the tropical hypersurface will obey the min-convention.
+# If M=max, the tropical hypersurface will obey the max-convention.
+# If coefficient ring has a valuation, the tropical hypersurface will be constructed with respect to it.
+# If coefficient ring has no valuation, the tropical hypersurface will be constructed with respect to the trivial valuation.
+# The function is the same as TropicalHypersurface{M}(f).
+
+# # Examples
+# julia> K = PadicField(7, 2)
+
+# julia> Kxy, (x,y) = K["x", "y"]
+
+# julia> f = 7*x+y+49
+
+# julia> tropical_variety(f,min)
+
+# julia> tropical_variety(f,max)
+# """
+# function tropical_variety(f::AbstractAlgebra.Generic.MPoly{<:RingElement}, M::Union{typeof(min),typeof(max)})
+#     return TropicalHypersurface{M}(f)
+# end
+
+
+
+###
+# 3. Basic properties
+# -------------------
+###
+
+# todo: add examples for varieties, curves and linear spaces
+@doc Markdown.doc"""
+    dual_subdivision(TH::TropicalHypersurface{M, EMB})
+
+Returns the dual subdivision of `TH` if it is embedded. Returns error otherwise
+
+# Examples
+A tropical hypersurface in RR^n is always of dimension n-1
+```jldoctest
+julia> T = tropical_semiring(min);
+
+julia> Txy,(x,y) = T["x","y"];
+
+julia> f = x+y+1;
+
+julia> tropicalLine = TropicalHypersurface(f);
+
+julia> dual_subdivision(tropicalLine)
+A subdivision of points in ambient dimension 3
+```
+"""
+function dual_subdivision(TH::TropicalHypersurface{M,EMB}) where {M,EMB}
+    # not sure whether it makes sense to support abstract tropical hypersurfaces, but it can't hurt to check
+    if !EMB
+        error("tropical hypersurface not embedded")
+    end
+
+    return SubdivisionOfPoints(pm_object(TH).DUAL_SUBDIVISION)
+end
+export dual_subdivision
+
+
+@doc Markdown.doc"""
+    polynomial(TH::TropicalHypersurface{M, EMB})
+
+Returns the tropical polynomial of `TH` if it is embedded. Returns error otherwise
```

Errors are not returned, they are thrown or raised.